collector/os_release.go

```diff
@@ -0,0 +1,179 @@
+				"variant", "variant_id", "version", "version_codename", "version_id"}, nil,
+		),
+		versionDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "os", "version"),
+			"Metric containing the major.minor part of the OS version.",
+			nil, nil,
+		),
+	}, nil
+}
+
+func parseOSRelease(r io.Reader) (*osRelease, error) {
+	env, err := envparse.Parse(r)
+	return &osRelease{
+		Name:            env["NAME"],
+		ID:              env["ID"],
+		IDLike:          env["ID_LIKE"],
+		PrettyName:      env["PRETTY_NAME"],
+		Variant:         env["VARIANT"],
+		VariantID:       env["VARIANT_ID"],
+		Version:         env["VERSION"],
+		VersionID:       env["VERSION_ID"],
+		VersionCodename: env["VERSION_CODENAME"],
+		BuildID:         env["BUILD_ID"],
+		ImageID:         env["IMAGE_ID"],
+		ImageVersion:    env["IMAGE_VERSION"],
+	}, err
+}
+
+func (c *osReleaseCollector) updateIfChanged(f *os.File, mtime *time.Time) error {
+	stat, err := f.Stat()
+	if err != nil {
+		return err
+	}
+
+	t := stat.ModTime()
+	if t != *mtime {
+		level.Debug(c.logger).Log("msg", "file modification time has changed",
+			"file", f.Name(), "old_value", mtime, "new_value", t)
+		*mtime = t
+
+		// Acquire a lock to update the osReleaseCollector struct.
+		c.osMutex.Lock()
+		defer c.osMutex.Unlock()
+
+		c.os, err = parseOSRelease(f)
+		if err != nil {
+			return err
+		}
+		majorMinor := versionRegex.FindString(c.os.VersionID)
+		if majorMinor != "" {
+			c.version, err = strconv.ParseFloat(majorMinor, 64)
+			if err != nil {
+				return err
+			}
+		} else {
+			c.version = 0
+		}
+	}
+	return nil
+}
+
+func (c *osReleaseCollector) UpdateStruct() error {
+	etcOSReleaseFile, err := os.Open(*rootfsPath + etcOSRelease)
+	if err == nil {
+		defer etcOSReleaseFile.Close()
+		return c.updateIfChanged(etcOSReleaseFile, &c.mtimeEtcOSRelease)
+	}
+
+	if !errors.Is(err, os.ErrNotExist) {
+		return err
+	}
+	c.mtimeEtcOSRelease = time.Time{}
+
+	// Fall back to /usr/lib/os-release if /etc/os-release is not present
+	usrLibOSReleaseFile, err := os.Open(*rootfsPath + usrLibOSRelease)
+	if err == nil {
+		defer usrLibOSReleaseFile.Close()
+		return c.updateIfChanged(usrLibOSReleaseFile, &c.mtimeUsrLibOSRelease)
+	}
+
+	if !errors.Is(err, os.ErrNotExist) {
+		return err
+	}
+	return fmt.Errorf("Neither %s nor %s exists: %w", etcOSReleaseFile.Name(), usrLibOSReleaseFile.Name(), err)
+}
+
+func (c *osReleaseCollector) Update(ch chan<- prometheus.Metric) error {
+	err := c.UpdateStruct()
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			level.Debug(c.logger).Log("msg", "no os-release file found, skipping")
+			return ErrNoData
+		}
+		return err
+	}
+	ch <- prometheus.MustNewConstMetric(c.infoDesc, prometheus.GaugeValue, 1.0,
+		c.os.BuildID, c.os.ID, c.os.IDLike, c.os.ImageID, c.os.ImageVersion, c.os.Name, c.os.PrettyName,
+		c.os.Variant, c.os.VariantID, c.os.Version, c.os.VersionCodename, c.os.VersionID)
+	if c.version > 0 {
```

Comment:
Not sure how useful this version metric is. Unless there is a good reason, I'd drop that and the parsing of the version

Reply:
In the past we had many questions which include checking for OS Release version newer than X or older than X. In that case it is simpler to write for example:
```
node_os_version{id="debian"} < 10
```
Instead of:
```
node_os_version{id="debian",version_id="9"} or node_os_version{id="debian",version_id="8"} or node_os_version{id="debian",version_id="7"} or node_os_version{id="debian",version_id="6"}
```